File: front/web/lib/goal-piloting.ts
/**
 * Goal piloting — the Workbench /goal-piloting CONTRACT TYPES (AIDOS step S66).
 *
 * S66 (ROADMAP-app-builder, KRD §56–§59, §63 ①): the UI-piloted /goal. A human PORTEUR
 * D'AUTORITÉ (S63) opens a goal from a grilled idea — the engine PROPOSES a DRAFT ChangeSet
 * (spec_delta + mirror_delta atomically) and computes the LIVE red set; the close gate is the
 * NON-GAMEABLE stop (red set→green ∧ prior intact ∧ mutation ≥ floor ∧ no monster).
 *
 * S59 CUTOVER (ADR 0092 — the Go engine is the SINGLE live source). This module USED to be the
 * DECLARED TWIN of the Go package back/runtime/goalpiloting — it re-implemented the actor gate, the
 * open, the close gate and the red-set sort in TS, and the panel read it as the live source. Those
 * READ functions are GONE: the goal-open, the non-gameable close and the red-set sort are now read
 * LIVE from the Go goal-piloting MCP server through the passerelle
 * (app/goal-piloting/actions.ts → readVia(scope, "goal_pilot_open" / "goal_pilot_close", …),
 * decoded by app/goal-piloting/live.ts). The demo fallback is the inert fixture lib/goal-piloting-data.
 *
 * WHAT REMAINS HERE (no twin logic): the CONTRACT TYPES the decoder fills + the panel renders
 * (PilotBlock, Delta, PilotOpenInput, PilotResult, and the re-exported RealActor / Budgets / Goal /
 * StopInput), plus the two canonical FR refusal constants (IDEA_WITHOUT_MIRROR / NO_RED_SET) the Go
 * engine returns verbatim — they document the contract the openDecoder maps, never a second
 * computation. The PLACEHOLDER_ACTOR / GOAL_STILL_RED refusals now come from the Go engine, decoded
 * by live.ts (no TS re-implementation).
 *
 * THE WALL (CLAUDE.md §2/§7): the live tools PROPOSE (a DRAFT ChangeSet) and VERDICT (the
 * non-gameable close) — they never APPLY a ChangeSet nor stamp CLOSED; the screen writes no Kernel.
 *
 * DETERMINISM-FIRST (CLAUDE.md §6/§8): no logic lives here anymore — the Go engine is authoritative;
 * the decoder (live.ts) is the only pure projection of its wire contract; "done" is computed by the
 * Go non-gameable stop, never declared.
 */

export type { RealActor } from "./authority-binding";
export type { Budgets, Goal, StopInput } from "./goal";

/** The unified, actionable refusal surfaced to the screen — the code carried verbatim. */
export interface PilotBlock {
	code: string;
	severity: string;
	explanation: string;
	howToFix: string[];
}

/** A delta the idea proposes (twin of changeset.Delta — only the fields the screen needs). */
export interface Delta {
	kind: string;
	target: string;
}

/** The screen-facing open input shape (twin of goalpiloting.OpenInput, the projected subset). */
export interface PilotOpenInput {
	ideaId: string;
	specDelta: Delta;
	/** Absent ⇒ a vœu (refused IDEA_WITHOUT_MIRROR). */
	mirrorDelta?: Delta;
	parentPhase: string;
	/** The LIVE red set the engine derived (S22 Impact) — the worklist. Non-empty for a real goal. */
	redSet: string[];
	budgets: import("./goal").Budgets;
}

/** What the screen receives when an authority-bearing human opens a goal (twin of PilotResult). */
export interface PilotResult {
	actor: import("./authority-binding").RealActor;
	goal: import("./goal").Goal;
}

/**
 * The canonical IDEA_WITHOUT_MIRROR refusal (the Go blockreason.For prose). KEPT as the documented
 * contract the openDecoder maps a refusal onto — NOT a second computation (the Go engine returns it).
 */
export const IDEA_WITHOUT_MIRROR: PilotBlock = {
	code: "IDEA_WITHOUT_MIRROR",
	severity: "blocking",
	explanation:
		"Refus de l'ouverture du goal : l'idée ne porte AUCUN mirror_delta. Une vérité sans miroir est un vœu (un monstre, §57/LIVRE XX) — le moteur de goal ne promeut jamais une idée sans son miroir. Aucun ChangeSet n'est ouvert.",
	howToFix: [
		"draft_mirror_for_idea : écrivez d'abord le miroir (Gherkin / propriété / fixture) qui prouve l'idée, puis rouvrez le /goal.",
		"idea → mirror → /goal : la seule porte légitime vers la vérité passe par le miroir.",
	],
};

/**
 * The canonical NO_RED_SET refusal (a green test is not a goal, §56). KEPT as the documented
 * contract the openDecoder maps a refusal onto — NOT a second computation (the Go engine returns it).
 */
export const NO_RED_SET: PilotBlock = {
	code: "NO_RED_SET",
	severity: "blocking",
	explanation:
		"Refus de l'ouverture du goal : le set rouge dérivé est VIDE. Un test déjà vert n'est pas un goal (§56 : le set rouge EST la todo-list) — il n'y a rien à fermer. Aucun goal n'est ouvert.",
	howToFix: [
		"bump_a_source : un goal naît quand un changement rougit au moins un miroir porteur.",
		"verify the red wave : vérifiez que le miroir reflète une source dont la tête a bougé (S22).",
	],
};

File: back/runtime/goalpiloting/goalpiloting.go
// Package goalpiloting is the AIDOS Runtime engine of S66 — the UI-PILOTED /goal.
// It is the surface a human PORTEUR D'AUTORITÉ (S63) drives from the Workbench to open
// a goal from a grilled idea WITHOUT ever writing the kernel (ROADMAP-app-builder S66,
// KRD §56–§59, §63 ①).
//
// WHAT IT ADDS over S29 (back/runtime/goal). S29 owns the bare, pure goal machinery —
// OpenGoal (DRAFT ChangeSet + derived red set) and IsClosed (the non-gameable stop).
// This package adds the genuinely-product concern S66 demands: a goal can only be
// PILOTED by a REAL acting human carrying authority (S63's RealActor / RequireRealActor),
// never a placeholder, never the agent. The acting human is the first gate; OpenGoal is
// the second; the four-condition stop is the close gate. Each gate DEFERS to its owner —
// nothing here re-implements a transition, a red-set derivation, or a close condition.
//
// THE WALL (CLAUDE.md §2). This package writes NOTHING. PilotOpenGoal returns a DRAFT
// ChangeSet PROPOSAL (a value) — the screen proposes it; the aidos CLI role persists it
// via the changeset door (S20) under human approval, never the agent. PilotCloseGoal is
// a pure predicate — it never stamps CLOSED. The screen never reaches the kernel.
//
// DETERMINISM-FIRST (CLAUDE.md §6/§8). Every function here is PURE and TOTAL — no clock,
// no rng, no I/O, never panics. The authority gate, the open and the close are all
// deterministic; the LLM never enters. Same input → same output; the reproducibility
// mirror goalpiloting_property_test.go pins it.
//
// FORWARD-DEPENDENCY (documented OpenQuestion, ROADMAP note 3). S63 (authoritybinding) is
// built, so PilotOpenGoal takes a REAL RealActor and the actor-gate is live. The full
// AuthorityGraph admission (role-floor approval over a scope domain) lands at S85's
// approval gate; until then S66 enforces the FIRST authority gate — the actor must be a
// real human (RequireRealActor), never a placeholder/agent. Scope-domain approval is
// annexed for S85, never silently faked here.
package goalpiloting

import (
	"github.com/steph-frtech/aidos/back/runtime/authoritybinding"
	"github.com/steph-frtech/aidos/back/runtime/blockreason"
	"github.com/steph-frtech/aidos/back/runtime/goal"
)

// Re-exports so a caller expresses the whole flow in goalpiloting terms while every type
// stays owned by its origin package (aliases — NOT forks; the same structs, the same code).

// Actor is the acting human a goal-open is attributed to (S63). A goal can never be opened
// by nobody: PilotOpenGoal refuses the empty/blank/placeholder actor.
type Actor = authoritybinding.RealActor

// OpenInput is S29's pure input to OpenGoal (the idea, the parent phase, the S22 red-wave
// inputs, the budgets). Reused verbatim.
type OpenInput = goal.OpenInput

// Goal is S29's promotion record (the idea_ref, the DRAFT ChangeSet, the red set, the
// computed status, the budgets). Reused verbatim.
type Goal = goal.Goal

// StopInput is S29's pure input to the non-gameable close predicate (the live sensors, the
// prior-green state, the mutation score + its declared floor, the monster findings).
type StopInput = goal.StopInput

// PilotBlock is the S66 unified, actionable refusal surfaced to the screen. It carries the
// refusal CODE VERBATIM (a string) from whichever gate refused — S63's actor codes
// (PLACEHOLDER_ACTOR / INSUFFICIENT_AUTHORITY, an S63-local BlockCode) or S29's open/close
// codes (IDEA_WITHOUT_MIRROR / NO_RED_SET / GOAL_STILL_RED, a blockreason.Code). Carrying
// the string AVOIDS inventing a member in either CLOSED enum (CLAUDE.md §9): the code is
// copied, never coined. The how_to_fix is carried verbatim so the wall always names the door.
type PilotBlock struct {
	Code        string   `json:"code"`
	Severity    string   `json:"severity"`
	Explanation string   `json:"explanation"`
	HowToFix    []string `json:"how_to_fix"`
}

func (b *PilotBlock) Error() string { return b.Code + ": " + b.Explanation }

// PilotResult is what the screen RECEIVES when an authority-bearing human opens a goal: the
// acting actor (carried for provenance + display) and the OPEN Goal carrying its DRAFT
// ChangeSet proposal + the live red set. It is a VALUE — nothing is persisted. The actor's
// real-actor provenance is what the changeset door (S20) records when the human approves.
type PilotResult struct {
	// Actor is the real acting human the goal-open is attributed to (S63). Carried so the
	// screen can render it and the changeset door can record real-actor provenance.
	Actor Actor `json:"actor"`
	// Goal is the OPEN goal S29 produced: the DRAFT ChangeSet proposal + the live red set.
	Goal Goal `json:"goal"`
}

// PilotOpenGoal is the S66 UI-piloted goal-open. It runs TWO gates in order and DEFERS each
// transition to its owner:
//
//  1. THE ACTOR GATE (S63): the acting human must be a REAL actor (not empty/blank, not a
//     declared placeholder, not "agent"). A placeholder is refused with PLACEHOLDER_ACTOR
//     — a goal-open (a step toward truth) is never attributed to nobody (the wall, §2).
//  2. THE OPEN (S29): OpenGoal opens a DRAFT ChangeSet wrapping the idea's spec_delta +
//     mirror_delta atomically and DERIVES the live red set via S22's Impact. A mirror-less
//     idea is refused (IDEA_WITHOUT_MIRROR); an empty red set is refused (NO_RED_SET).
//
// On success it returns a PilotResult carrying the actor + the OPEN goal (a DRAFT ChangeSet
// PROPOSAL + the live red set). It WRITES NOTHING. PURE: no DB, no clock, no rng.
//
// The returned PilotBlock carries the refusal code verbatim from whichever gate refused —
// the S63 actor refusal or S29's own open refusal — both actionable (code + how_to_fix).
func PilotOpenGoal(actor Actor, in OpenInput) (PilotResult, *PilotBlock) {
	// (1) The actor gate — a real human, never a placeholder, never the agent (S63, the wall).
	if abr := authoritybinding.RequireRealActor(actor); abr != nil {
		b := fromActorBlock(abr)
		return PilotResult{}, &b
	}
	// (2) The open — S29 owns the DRAFT ChangeSet + the live red-set derivation.
	g, gbr := goal.OpenGoal(in)
	if gbr != nil {
		b := fromGoalBlock(gbr)
		return PilotResult{}, &b
	}
	return PilotResult{Actor: actor, Goal: g}, nil
}

// PilotCloseGoal is the S66 UI-piloted close attempt. It DEFERS entirely to S29's
// non-gameable stop (goal.IsClosed): a goal closes iff
//
//	red set → green ∧ prior green intact ∧ mutation ≥ floor ∧ no monster.
//
// It returns the actionable GOAL_STILL_RED refusal while any condition fails (nil when the
// goal IS closeable). It NEVER stamps the goal CLOSED — that truth-write stays the aidos CLI
// role via S20's commit-gate. PURE and TOTAL; takes NO agent-confidence input (there is no
// such field) — "done" is computed, never declared (§8). This is the screen's close gate.
func PilotCloseGoal(g Goal, in StopInput) *PilotBlock {
	gbr := goal.CloseBlockReason(g, in)
	if gbr == nil {
		return nil
	}
	b := fromGoalBlock(gbr)
	return &b
}

// CanClose reports whether the four-condition non-gameable stop holds for a goal — a small
// convenience for the screen's live close indicator. DEFERS to goal.IsClosed. PURE.
func CanClose(g Goal, in StopInput) bool {
	return goal.IsClosed(g, in)
}

// LiveRedSet returns the goal's live red set in stable sorted order — the worklist the panel
// renders (each entry a failing mirror red → green). DEFERS to goal.RedSetSorted. PURE.
func LiveRedSet(g Goal) []string {
	return goal.RedSetSorted(g)
}

// fromActorBlock carries S63's actor refusal verbatim into the S66 PilotBlock — the code
// (PLACEHOLDER_ACTOR / INSUFFICIENT_AUTHORITY) stringified, never coined (CLAUDE.md §9). Pure.
func fromActorBlock(abr *authoritybinding.BlockReason) PilotBlock {
	return PilotBlock{
		Code:        string(abr.Code),
		Severity:    abr.Severity,
		Explanation: abr.Explanation,
		HowToFix:    abr.HowToFix,
	}
}

// fromGoalBlock carries S29's open/close refusal verbatim into the S66 PilotBlock — the code
// (IDEA_WITHOUT_MIRROR / NO_RED_SET / GOAL_STILL_RED) stringified, never coined. Pure.
func fromGoalBlock(gbr *blockreason.BlockReason) PilotBlock {
	return PilotBlock{
		Code:        string(gbr.Code),
		Severity:    string(gbr.Severity),
		Explanation: gbr.Explanation,
		HowToFix:    gbr.HowToFix,
	}
}
